sqlglot/optimizer/simplify.py

```diff
@@ -369,6 +370,49 @@ def absorb_and_eliminate(expression, root=True):
     return expression
 
 
+def propagate_constants(expression, root=True):
+    """
+    Propagate constants for conjunctions in DNF:
+
+    SELECT * FROM t WHERE a = b AND b = 5 becomes
+    SELECT * FROM t WHERE a = 5 AND b = 5
+
+    Reference: https://www.sqlite.org/optoverview.html
+    """
+    from sqlglot.optimizer.normalize import normalized
+
+    if (
+        isinstance(expression, exp.And)
+        and (root or not expression.same_parent)
+        and normalized(expression, dnf=True)
+    ):
+        constant_mapping: t.Dict[exp.Column, t.Tuple[int, exp.Literal]] = {}
+        for eq in expression.find_all(exp.EQ):
+            l, r = eq.left, eq.right
+
+            if isinstance(l, exp.Column) and isinstance(r, exp.Literal):
+                pass
+            elif isinstance(r, exp.Column) and isinstance(l, exp.Literal):
+                l, r = r, l
+            else:
+                continue
+
+            constant_mapping[l] = (id(l), r)
+
+        if constant_mapping:
+            for column in expression.find_all(exp.Column):
+                parent = column.parent
+                column_id, constant = constant_mapping.get(column) or (None, None)
+                if (
+                    column_id is not None
+                    and id(column) != column_id
+                    and not (isinstance(parent, exp.Is) and isinstance(parent.expression, exp.Null))
+                ):
+                    column.replace(constant.copy())
```

Comment:
Is replacement right? That takes it a step further than the sqlite rule, which says it just adds the extra predicate.

Do any planners take advantage of column equality in join conditions?

Reply:
Doesn't SQLite also replace the variable? Based on the example I thought it does

Reply:
The example doesn’t show replacement. I also wonder if there is different behavior on join conditions.

Reply:
Yeah, will dive deeper. Also a good catch, thanks!

Reply:
It seems that there are some gnarly cases that are tricky to take care of - mostly because of coercions. Interestingly, SQLite doesn't implement this transformation for `ON` clauses of `LEFT`, `RIGHT` joins:

```c
      ...
      /* Do not propagate constants on any ON clause if there is a
      ** RIGHT JOIN anywhere in the query */
      x.mExcludeOn = EP_InnerON | EP_OuterON;
    }else{
      /* Do not propagate constants through the ON clause of a LEFT JOIN */
      x.mExcludeOn = EP_OuterON;
      ...
```

Some documentation on SQLite's implementation can be found [here](https://github.com/tobymao/sqlglot/assets/46752250/e97b5a63-f0cb-46f2-a47d-3f69bfe58376) (screenshot of their [source code](https://sqlite.org/src/doc/trunk/README.md)).

On the other hand, MariaDB [mentions](https://mariadb.com/kb/en/equality-propagation-optimization/) this transformation in the context of `WHERE` clauses, but towards the end of the page they also have this (note the remark about `ON` expressions):

<img width="753" alt="Screenshot 2023-10-09 at 5 46 56 AM" src="https://github.com/tobymao/sqlglot/assets/46752250/945f3f47-1f3e-4805-aed5-8aa1b2575efb">

Reply:
Regarding

> The example doesn’t show replacement

It's now clear that SQLite does indeed replace the variable with the value:

<img width="547" alt="Screenshot 2023-10-09 at 6 03 43 AM" src="https://github.com/tobymao/sqlglot/assets/46752250/a750481b-8062-4b97-9634-181eb941a1c8">